# Algoritmos de Optimización - Proyecto de Programacion

Nombre: ***Oscar Patricio Cuenca Moreno*** <br>

[GitHub](https://github.com/HikariJY/03MIAR_04_A_2024-25_Algoritmos-de-Optimizacion/tree/main/Actividades/Seminario01) Pendiente cambiar por el mio

# Enunciado del problema

## Problema 3 -- Combinar Cifras y Operaciones

- El problema consiste en analizar el siguiente problema y diseñar un algoritmo que lo resuelva.
- Disponemos de las 9 cifras del 1 al 9 (excluimos el cero) y los 4 signos básicos de las operacinoes fundamentales.

 - Suma (+).
 - Resta (-).
 - Multiplicación (*)
 - División (/).

- Debemos combinarlos alternativamente sin repetir ninguno de ellos para obtener una cantidad dada.

> Un ejemplo, seria para obtener el 4: <br>
> $4+2-6/3*1=4$

Debe analizarse el problema para encontrar todos los valores enteros posibles planteando las siguientes cuestiones:
- ¿Qué valor máximo y mínimo se pueden obtener según las condiciones del problema?
- ¿Es posible encontrar todos los valores enteros posibles entre dicho mínimo y máximo ?
- Nota: Es posible usar la función de python “eval” para evaluar una expresión

# Respuestas a las preguntas
## ¿Qué valor máximo y mínimo se pueden obtener según las condiciones del problema?

In [11]:
import itertools as it  # Importamos la librería itertools para poder generar las permutaciones

def evaluate_expression(numbers, operators):
    """
    Función que usaremos para evaluar todas las combinaciones posibles de números y operadores
    para poder encontrar el valor mínimo y máximo, asegurándonos que las divisiones sean enteras.
    Además, almacenaremos todos los resultados enteros encontrados.
    """
    r_min, r_max = float('inf'), float('-inf')  # Inicializamos los valores mínimo y máximo
    results = set()  # Creamos un conjunto para guardar todos los resultados enteros únicos

    # Generamos todas las permutaciones posibles de 5 números y 4 operadores
    for nums in it.permutations(numbers, 5):  # Permutaciones de números
        for ops in it.permutations(operators, 4):  # Permutaciones de operadores
            try:
                # Verificamos que la división sea entera antes de evaluar la expresión
                if nums[ops.index('/')] % nums[ops.index('/') + 1] == 0:
                    # Construimos la expresión matemática como una cadena
                    expression = "".join(f"{nums[i]}{ops[i]}" for i in range(4)) + str(nums[4])
                    # Evaluamos la expresión con eval y verificamos si el resultado es entero
                    result = eval(expression)
                    if result.is_integer():  # Solo contamos resultados enteros
                        result = int(result)  # Convertimos el resultado a entero
                        results.add(result)  # Agregamos el resultado al conjunto
                        # Actualizamos el mínimo y el máximo
                        r_min, r_max = min(r_min, result), max(r_max, result)
            except ZeroDivisionError:  # Si ocurre división por cero, continuamos sin detener el programa
                continue

    return r_min, r_max, results  # Devuelvemos el mínimo, máximo y el conjunto de resultados enteros

# Datos de entrada
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9]  # Lista de números disponibles
operators = ['+', '-', '*', '/']  # Lista de operadores disponibles

# Hacemos el llamado a la función para calcular el mínimo, máximo y todos los resultados enteros
r_min, r_max, all_results = evaluate_expression(numbers, operators)

# Calculamos el rango esperado de valores entre el mínimo y el máximo
expected_values = set(range(r_min, r_max + 1))  # Creamos un conjunto con todos los valores esperados
missing_values = expected_values - all_results  # Calculamos los valores faltantes comparando conjuntos

# Mostramos los resultados obtenidos
print(f"El valor mínimo es: {r_min}.")
print(f"El valor máximo es: {r_max}.")
print(f"Total de valores enteros obtenidos: {len(all_results)}")

# Verificamos si faltan valores
if not missing_values:
    print("Todos los valores enteros entre el mínimo y el máximo son obtenibles.")
else:
    print("Faltan los siguientes valores enteros:")
    print(sorted(missing_values))

El valor mínimo es: -69.
El valor máximo es: 77.
Total de valores enteros obtenidos: 147
Todos los valores enteros entre el mínimo y el máximo son obtenibles.


## ¿Es posible encontrar todos los valores enteros posibles entre dicho mínimo y máximo ?

Si es posible, obteniendo los 147 valores.

## ¿Cuantas posibilidades hay sin tener en cuenta las restricciones?

Dado un conjunto de números \( numbers = \{1, 2, 3, 4, 5, 6, 7, 8, 9\} \) y un conjunto de operadores \( operators = \{+, -, *, /\} \), si no imponemos restricciones sobre la repetición de números y operadores, se tienen:

- **5 posiciones** para los números, donde cada posición puede tomar **cualquiera de los 9 valores**.
- **4 posiciones** para los operadores, donde cada posición puede tomar **cualquiera de los 4 valores**.

El total de combinaciones posibles se calcula como:
\[
P = 9 \times 9 \times 9 \times 9 \times 9 \times 4 \times 4 \times 4 \times 4 = 9^5 \times 4^4
\]

Resolviendo:
\[
P = 15116544
\]

### Resultado:
**El número total de posibilidades sin tener en cuenta las restricciones es 15,116,544.**

In [12]:
# Cálculo del total de combinaciones sin las restricciones
def total_possibilities():
    """
    Calculamos el número total de combinaciones posibles usando:
    - 5 posiciones para números (conjunto de 9 elementos, permitiendo repetición).
    - 4 posiciones para operadores (conjunto de 4 elementos, permitiendo repetición).

    El cálculo sigue la fórmula: 9^5 * 4^4.
    """
    numbers_pos = 9 ** 5  # 9 opciones por cada una de las 5 posiciones de números
    operators_pos = 4 ** 4  # 4 opciones por cada una de las 4 posiciones de operadores

    total = numbers_pos * operators_pos  # Combinaciones totales
    return total

# Llamada a la función y muestra del resultado
result = total_possibilities()
print(f"El número total de posibilidades sin restricciones es: {result}")

El número total de posibilidades sin restricciones es: 15116544


## ¿Cuántas posibilidades hay teniendo en cuenta todas las restricciones?

Al considerar las restricciones del problema, debemos garantizar lo siguiente:

1. **Restricciones en los números**:
   - Seleccionamos **5 números únicos** del conjunto \( \{1, 2, 3, 4, 5, 6, 7, 8, 9\} \).  
   Esto corresponde a una **permutación de 9 elementos tomados de 5**:
   \[
   P(9, 5) = \frac{9!}{(9-5)!} = 9 \cdot 8 \cdot 7 \cdot 6 \cdot 5 = 15120
   \]

2. **Restricciones en los operadores**:
   - Seleccionamos **4 operadores únicos** del conjunto \( \{+, -, *, /\} \) y los colocamos en 4 posiciones.  
   Esto corresponde a una **permutación de 4 elementos**:
   \[
   P(4, 4) = 4! = 4 \cdot 3 \cdot 2 \cdot 1 = 24
   \]

3. **Restricción de divisiones enteras**:
   - La posición del operador división `/` debe cumplir que el número anterior sea divisible entre el siguiente número sin residuo.  
   Esto impone restricciones adicionales que deben verificarse en cada combinación.

---

### Cálculo total preliminar:
La cantidad inicial de posibilidades, sin contar aún las divisiones inválidas, es:
\[
\text{Total sin divisiones inválidas} = P(9, 5) \cdot P(4, 4) = 15120 \cdot 24 = 362880
\]

Para obtener las combinaciones **válidas**, debemos filtrar únicamente las expresiones donde la división sea entera.

---

### Resultado final:
La cantidad exacta de combinaciones válidas se debe obtener programáticamente al **filtrar las restricciones** en un algoritmo, asegurando que:
1. La división sea entera cuando se use el operador `/`.
2. Se generen todas las permutaciones posibles de números y operadores.

El resultado final será un subconjunto del total **362,880 combinaciones** iniciales.

In [16]:
import itertools as it  # Importamos itertools para generar permutaciones

def possibilities_with_and_without_restrictions(numbers, operators):
    """
    Calcula las combinaciones válidas en dos casos:
    - Caso 1: Sin restricciones (todas las combinaciones se cuentan como válidas).
    - Caso 2: Con restricciones, verificando que la división sea entera cuando aparece '/'.
    """
    total_no_restrictions = 0  # Contador para combinaciones sin restricciones
    total_with_restrictions = 0  # Contador para combinaciones con restricciones

    # Generamos todas las permutaciones de 5 números únicos
    for nums in it.permutations(numbers, 5):
        # Generamos todas las permutaciones de 4 operadores únicos
        for ops in it.permutations(operators, 4):
            # Caso 1: Contamos todas las combinaciones sin restricciones
            total_no_restrictions += 1

            try:
                # Caso 2: Aplicamos las restricciones si hay división
                if '/' in ops:
                    div_index = ops.index('/')  # Encuentro la posición del operador '/'
                    if nums[div_index] % nums[div_index + 1] != 0:  # Verificamos división entera
                        continue  # Si no cumple, no contamos esta combinación
                # Evaluamos la expresión
                expression = "".join(f"{nums[i]}{ops[i]}" for i in range(4)) + str(nums[4])
                eval(expression)  # Evaluamos la expresión para confirmar validez
                total_with_restrictions += 1
            except ZeroDivisionError:
                continue  # Ignoramos divisiones por cero

    return total_no_restrictions, total_with_restrictions  # Devolvemos ambos resultados

# Datos de entrada
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9]  # Lista de números disponibles
operators = ['+', '-', '*', '/']  # Lista de operadores disponibles

# Llamada a la función
no_restrictions, with_restrictions = possibilities_with_and_without_restrictions(numbers, operators)

# Mostramos los resultados
print(f"Combinaciones sin restricciones: {no_restrictions}")
print(f"Combinaciones con restricciones (división entera): {with_restrictions}")

Combinaciones sin restricciones: 362880
Combinaciones con restricciones (división entera): 70560


## Modelo para el espacio de soluciones
## ¿Cual es la estructura de datos que mejor se adapta al problema? Argumentalo.(Es posible que hayas elegido una al principio y veas la necesidad de cambiar, arguentalo)

Para resolver el problema mediante Fuerza Bruta, resulta útil usar una lista simple que almacene los números disponibles y otra lista para los operadores. Estas listas son eficientes para acceder a los elementos y permiten realizar las permutaciones necesarias fácilmente. Ademas para la construcción y evaluación de la expresión matemática, una variable de tipo cadena de texto (str) es adecuada, ya que se puede concatenar dinámicamente y evaluar utilizando la función eval de Python para obtener el resultado.
No ha sido necesario realizar cambios en las estructuras de datos iniciales, ya que las listas y las cadenas de texto se ajustan bien a las operaciones necesarias para resolver el problema.

## Según el modelo para el espacio de soluciones
## ¿Cual es la función objetivo?
La función objetivo consiste en ejecutar un algoritmo que genere todas las combinaciones posibles de números y operadores, buscando que el resultado de la ecuación evaluada coincida con un valor dado. La diferencia entre el valor obtenido y el esperado actúa como una medida de error, permitiendo determinar qué soluciones son aceptables y mostrando la expresión matemática correspondiente.
## ¿Es un problema de maximización o minimización?
Se trata de un problema de minimización, ya que el objetivo es reducir al máximo el error entre el resultado calculado y el valor deseado. Sin embargo, debido a la naturaleza discreta de las operaciones, no es apropiado utilizar métodos como el descenso de gradiente para resolverlo.

## Diseña un algoritmo para resolver el problema por fuerza bruta

In [42]:
import itertools as it  # Importo itertools para generar permutaciones

def brute_force(target, numbers, operators, verbose=True):
    """
    Algoritmo de Fuerza Bruta para poder encontrar las expresiones que evalúan un valor objetivo.
    Parámetros:
    - target: Valor que queremos obtener.
    - numbers: Lista de números disponibles.
    - operators: Lista de operadores disponibles.
    - verbose: Flag para imprimir o no las soluciones encontradas.
    """
    # Generamos todas las combinaciones de 5 números únicos
    for numb in it.permutations(numbers, 5):
        # Generamos todas las combinaciones de 4 operadores únicos
        for oper in it.permutations(operators, 4):
            # Construimos la expresión matemática como una cadena
            expr = f"{numb[0]}{oper[0]}{numb[1]}{oper[1]}{numb[2]}{oper[2]}{numb[3]}{oper[3]}{numb[4]}"
            try:
                # Verificamos si la expresión coincide con el valor objetivo
                if eval(expr) == target:
                    if verbose:  # Si verbose es True, mostramos la solución
                        print(f"Solución encontrada: {expr} = {target}")
                    return True  # Devolvemos True porque encontré una solución válida
            except ZeroDivisionError:
                continue  # Ignoro divisiones por cero y paso a la siguiente combinación
    if verbose:  # Si no se encontró ninguna solución
        print(f"No se encontró una expresión para {target}")
    return False  # Devuelvo False porque no encontré una solución válida

# Datos de entrada
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9]  # Números disponibles
operators = ['+', '-', '*', '/']  # Operadores disponibles
target = 69  # Valor objetivo

# Llamo a la función y muestro el resultado
result = brute_force(target=target, numbers=numbers, operators=operators)

Solución encontrada: 2/1-5+8*9 = 69


## Calcula la complejidad del algoritmo por fuerza bruta

#### Calculo de complejidad:

1. **Permutaciones de números**:
   - Seleccionamos \( k = 5 \) números únicos de un conjunto de tamaño \( n \).
   - La cantidad de permutaciones es:
     \[
     P(n, k) = \frac{n!}{(n-k)!}
     \]
     Como \( k = 5 \), la fórmula se simplifica a:
     \[
     P(n, 5) = n \cdot (n-1) \cdot (n-2) \cdot (n-3) \cdot (n-4)
     \]
     La complejidad de esta parte es **O(n^5)**, porque multiplicamos \( n \) decreciendo 5 veces.

2. **Permutaciones de operadores**:
   - El número total de operadores (\( m = 4 \)) permanece constante, y se calculan todas las permutaciones:
     \[
     P(m, m) = m! = 4! = 24
     \]
   - Como \( m \) es constante, esta parte tiene complejidad **O(1)**.

3. **Construcción y evaluación de la expresión**:
   - La concatenación de números y operadores, así como la evaluación de la expresión, se realiza en tiempo constante **O(1)**.

---

#### Complejidad total:

Multiplicamos las permutaciones de números (\( O(n^5) \)) por las permutaciones de operadores (\( O(1) \)):

\[
O(T(n)) = O(n^5) \cdot O(1) = O(n^5)
\]

---

### Finalmente:

La **complejidad del algoritmo por fuerza bruta** en función de \( n \), el tamaño del conjunto de números, es:

\[
O(n^5)
\]

Esto asume que el número de operadores \( m \) es constante (\( m = 4 \)). Si \( m \) creciera, su complejidad también debería considerarse en la fórmula.

## Diseña un algoritmo que mejore la complejidad del algortimo por fuerza bruta. Argumenta porque crees que mejora el algoritmo por fuerza bruta

In [26]:
def branching_pruning(target, numbers=None, operators=None, expr='', idx=0, verbose=True):
    """
    Algoritmo de Ramificación y Poda para encontrar expresiones válidas que igualen el valor objetivo.
    Parámetros:
    - target: Valor objetivo que queremos encontrar.
    - numbers: Lista de números disponibles.
    - operators: Lista de operadores disponibles.
    - expr: Expresión matemática parcial.
    - idx: Índice actual en la construcción de la expresión.
    - verbose: Si True, imprime las soluciones encontradas.
    """
    if numbers is None:
        numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9]
    if operators is None:
        operators = ['+', '-', '*', '/']

    try:
        # Caso base: Si completamos la expresión (idx > 8), la evaluamos
        if idx > 8:
            if eval(expr) == target:  # Evaluación de la expresión completa
                if verbose:
                    print(f"{expr} = {target}")
                return True
            return False

        # Si el índice es PAR -> Añadimos un número
        if idx % 2 == 0:
            for num in numbers:
                new_expr = expr + str(num)  # Construimos la nueva expresión
                new_numbers = [x for x in numbers if x != num]  # Quitamos el número usado

                # Validación especial para divisiones
                if '/' in expr:
                    idx_div = expr.find('/')
                    prev_num = int(expr[idx_div - 1])  # Número anterior a la división

                    # Verificar divisiones exactas
                    if prev_num % num != 0:
                        continue  # Poda: descartamos la expresión si la división no es exacta

                    # Poda por error grande
                    if abs(eval(new_expr) - target) > 8:
                        continue

                # Llamada recursiva
                if branching_pruning(target, new_numbers, operators, new_expr, idx + 1, verbose):
                    return True

        # Si el índice es IMPAR -> Añadimos un operador
        else:
            for op in operators:
                new_expr = expr + op  # Añadimos el operador a la expresión
                new_operators = [x for x in operators if x != op]  # Quitamos el operador usado

                # Llamada recursiva
                if branching_pruning(target, numbers, new_operators, new_expr, idx + 1, verbose):
                    return True

    except ZeroDivisionError:  # Manejo seguro de divisiones por cero
        return False

    return False


# Llamada al algoritmo
target = 69
branching_pruning(target)

2+8*9-5/1 = 69


True

Este algoritmo mejora el enfoque de fuerza bruta al aplicar ramificación y poda para reducir el espacio de búsqueda. La ramificación organiza la construcción de la expresión alternando números y operadores, garantizando combinaciones válidas, mientras que la poda descarta ramas innecesarias de manera temprana al verificar que las divisiones sean exactas y eliminando aquellas expresiones parciales cuyo resultado se aleja demasiado del objetivo. Esto optimiza el tiempo de ejecución al evitar evaluaciones redundantes y garantiza que solo se consideren caminos potencialmente válidos.

### Calcula la complejidad del algoritmo

In [27]:
evaluated = 0  # Contador de combinaciones evaluadas
pruned = 0     # Contador de combinaciones podadas

def branching_pruning(target, numbers=None, operators=None, expr='', idx=0, verbose=True):
    global evaluated, pruned

    try:
        if idx > 8:  # Evaluamos la expresión completa
            evaluated += 1
            if eval(expr) == target:
                if verbose:
                    print(f"{expr} = {target}")
                return True
            return False

        if idx % 2 == 0:  # Añadir un número
            for num in numbers:
                new_expr = expr + str(num)
                new_numbers = [x for x in numbers if x != num]

                # Verificamos división exacta
                if '/' in expr:
                    idx_div = expr.find('/')
                    prev_num = int(expr[idx_div - 1])
                    if prev_num % num != 0:
                        pruned += 1
                        continue

                # Llamada recursiva
                if branching_pruning(target, new_numbers, operators, new_expr, idx + 1, verbose):
                    return True

        else:  # Añadir un operador
            for op in operators:
                new_expr = expr + op
                new_operators = [x for x in operators if x != op]

                # Llamada recursiva
                if branching_pruning(target, numbers, new_operators, new_expr, idx + 1, verbose):
                    return True

    except ZeroDivisionError:
        pruned += 1
        return False

    return False

# Datos de entrada
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9]
operators = ['+', '-', '*', '/']
target = 69

evaluated, pruned = 0, 0  # Reiniciamos los contadores
branching_pruning(target, numbers, operators)
print(f"Combinaciones evaluadas: {evaluated}")
print(f"Ramas podadas: {pruned}")

2+8*9-5/1 = 69
Combinaciones evaluadas: 1767
Ramas podadas: 15703


El espacio de búsqueda original en el enfoque de **fuerza bruta** tiene una complejidad de \( O(n^5) \), correspondiente a todas las combinaciones posibles de números y operadores.

Al aplicar **ramificación y poda**, el número de combinaciones evaluadas se reduce considerablemente. A partir de los resultados obtenidos:

- **Combinaciones totales (fuerza bruta):** \( 362880 \)
- **Combinaciones evaluadas:** \( 1767 \)
- **Proporción de poda:**  
  \[
  \text{Proporción} = \frac{1767}{362880} \approx 0.00487
  \]

Usando esta proporción para ajustar la complejidad, tenemos:
\[
n^k = 0.00487 \cdot n^5
\]
Tomando logaritmos en base \( n = 9 \), se obtiene:
\[
k = 5 + \frac{\log(0.00487)}{\log(9)} \approx 2.58
\]

---

### **Resultado**
La complejidad del algoritmo de **ramificación y poda** se reduce aproximadamente a:
\[
O(n^{2.58})
\]

Esto representa una **mejora significativa** respecto a la complejidad original de \( O(n^5) \) del enfoque de fuerza bruta.

### Según el problema (y tenga sentido), diseña un juego de datos de entrada aleatorios

In [47]:
import numpy as np
import random

def generar_juego_datos():
    """
    Genera un conjunto de datos aleatorios para el problema:
    - 5 números únicos del 1 al 9.
    - 4 operadores seleccionados aleatoriamente con repetición.
    - 10 valores objetivo aleatorios entre el mínimo (-69) y el máximo (77) conocidos.
    """
    # Generamos los 5 números únicos del 1 al 9
    numeros = [1,2,3,4,5,6,7,8,9]

    # Generamos 4 operadores aleatorios permitiendo repetición
    operadores = ['+', '-', '*', '/']

    # Generamos 10 valores objetivo aleatorios entre el mínimo y el máximo conocidos
    all_ints = np.arange(-69, 78)  # Rango de valores enteros posibles
    valores_objetivo = np.random.choice(all_ints, size=10, replace=False)  # 15 valores únicos

    # Mostrar los datos generados
    print("Juego de datos aleatorio generado:")
    print(f"Números: {numeros}")
    print(f"Operadores: {operadores}")
    print(f"Valores objetivo: {valores_objetivo.tolist()}")

    return numeros, operadores, valores_objetivo

# Generar y mostrar el juego de datos
numeros, operadores, valores_objetivo = generar_juego_datos()

Juego de datos aleatorio generado:
Números: [1, 2, 3, 4, 5, 6, 7, 8, 9]
Operadores: ['+', '-', '*', '/']
Valores objetivo: [16, -30, 31, -56, -60, -50, 3, 51, 28, -15]


### Aplica el algoritmo al juego de datos generado

In [48]:
import time

# Aplicación del Algoritmo de Fuerza Bruta
print("\n### Ejecución del Algoritmo de Fuerza Bruta ###")
start_time = time.time()  # Tiempo inicial

not_found = []  # Lista de valores no encontrados
for target in valores_objetivo:  # Recorremos cada valor objetivo generado
    res = brute_force(target=target, numbers=numeros.copy(), operators=operadores.copy(), verbose=True)
    if not res:
        not_found.append(target)

if not_found:
    print(f"No se encontraron todos los valores enteros: {not_found}.")
else:
    print("Se encontraron todos los valores enteros.")

execution_time = time.time() - start_time  # Tiempo total de ejecución
print(f"El tiempo de Fuerza Bruta fue: {execution_time:.2f} s.")

# Aplicación del Algoritmo de Ramificación y Poda
print("\n### Ejecución del Algoritmo de Ramificación y Poda ###")
start_time = time.time()  # Tiempo inicial

not_found = []  # Reiniciamos la lista de valores no encontrados
for target in valores_objetivo:  # Recorremos cada valor objetivo generado
    res = branching_pruning(target=target, numbers=numeros.copy(), operators=operadores.copy(), verbose=True)
    if not res:
        not_found.append(target)

if not_found:
    print(f"No se encontraron todos los valores enteros: {not_found}.")
else:
    print("Se encontraron todos los valores enteros.")

execution_time = time.time() - start_time  # Tiempo total de ejecución
print(f"El tiempo de Ramificación y Poda fue: {execution_time:.2f} s.")


### Ejecución del Algoritmo de Fuerza Bruta ###
Solución encontrada: 1-3+4/2*9 = 16
Solución encontrada: 1-5*7+8/2 = -30
Solución encontrada: 1+4*8-6/3 = 31
Solución encontrada: 2/1+5-7*9 = -56
Solución encontrada: 1+4/2-7*9 = -60
Solución encontrada: 2/1+4-7*8 = -50
Solución encontrada: 1-2+3/6*8 = 3
Solución encontrada: 1+6*9-8/2 = 51
Solución encontrada: 1+5*6-9/3 = 28
Solución encontrada: 1+2-6/3*9 = -15
Se encontraron todos los valores enteros.
El tiempo de Fuerza Bruta fue: 2.22 s.

### Ejecución del Algoritmo de Ramificación y Poda ###
1-3+9*4/2 = 16
1+5-9*8/2 = -30
1+4*8-6/3 = 31
2+5-7*9/1 = -56
1-7*9+4/2 = -60
2+4-7*8/1 = -50
1+6-2*8/4 = 3
1+6*9-8/2 = 51
1+5*6-9/3 = 28
1+2-6*9/3 = -15
Se encontraron todos los valores enteros.
El tiempo de Ramificación y Poda fue: 0.17 s.


### Enumera las referencias que has utilizado(si ha sido necesario) para llevar a cabo el trabajo

- Universidad Internacional de Valencia (VIU).  
  *Material de Algoritmos de Optimización*. 2024. Disponible en la plataforma Learn VIU.  
  [https://learn.universidadviu.com](https://learn.universidadviu.com)


### Describe brevemente las lineas de como crees que es posible avanzar en el estudio del problema. Ten en cuenta incluso posibles variaciones del problema y/o variaciones al alza del tamaño

**Optimización de algoritmos**: Incorporar técnicas como **memoización** o heurísticas para reducir el espacio de búsqueda.  
**Ampliación del problema**: Analizar el comportamiento al **aumentar el número de elementos** o introducir operadores adicionales como potencias (^) o módulo (%).  
**Metaheurísticas**: Explorar métodos como **algoritmos genéticos** para soluciones aproximadas en problemas de mayor tamaño.  
**Paralelización**: Adaptar los algoritmos a entornos **paralelos o distribuidos** para mejorar el rendimiento.  
**Variaciones del problema**: Permitir la **repetición de números y operadores** o aplicar restricciones adicionales.